content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,137 @@
+
+In cgroup v1, and prior to this feature, the container runtime never took into account and effectively ignored spec.containers[].resources.requests["memory"]. This is unlike CPU, in which the container runtime consider both requests and limits. Furthermore, memory actually can't be compressed in cgroup v1. Because there is no way to throttle memory usage, if a container goes past its memory limit it will be terminated by the kernel with an OOM (Out of Memory) kill.
+
+Fortunately, cgroup v2 brings a new design and implementation to achieve full protection on memory. The new feature relies on cgroups v2 which most current operating system releases for Linux already provide. With this experimental feature, [quality-of-service for pods and containers](/docs/tasks/configure-pod-container/quality-service-pod/) extends to cover not just CPU time but memory as well.
+
+## How does it work?
+Memory QoS uses the memory controller of cgroup v2 to guarantee memory resources in Kubernetes. Memory requests and limits of containers in pod are used to set specific interfaces `memory.min` and `memory.high` provided by the memory controller. When `memory.min` is set to memory requests, memory resources are reserved and never reclaimed by the kernel; this is how Memory QoS ensures the availability of memory for Kubernetes pods. And if memory limits are set in the container, this means that the system needs to limit container memory usage, Memory QoS uses `memory.high` to throttle workload approaching it's memory limit, ensuring that the system is not overwhelmed by instantaneous memory allocation.
+
+![](./memory-qos-cal.svg)
+
+The following table details the specific functions of these two parameters and how they correspond to Kubernetes container resources.
+
+<table>
+    <tr>
+        <th style="text-align:center">File</th>
+        <th style="text-align:center">Description</th>
+   </tr>
+   <tr>
+        <td>memory.min</td>
+        <td><code>memory.min</code> specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked.
+        <br>
+        <br>
+        <i>We map it to the container's memory request</i>
+        </td>
+   </tr>
+   <tr>
+       <td>memory.high</td>
+       <td><code>memory.high</code> is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. 
```

Comment:
> The default is max, ...

What does 'max' mean here?

Reply:
> memory.high
> - A read-write single value file which exists on non-root cgroups. The default is “max”.

The `max` is the file content of memory.high or memory.max when it is unlimited. The unlimited is so-called `noop` as well.

Reply:
It is probably fine to copy paste contents from somewhere else, so long as the license permits.
However, when reading this post, we need to explain what do you mean by saying "max".

The same applies to the previous paragraphs. It doesn't matter whether the contents were copied from somewhere.
What really matters is that the new blog should be useful, precise and concise.
It is a bad article if people cannot understand what you are talking about.

Reply:
@tengqm this is a blog post for an existing feature which is modified in v1.27. Hence, I copied the blog post from previous version and added my changes on top of it. However, it does make sense to have a concise blog. I have restructured the blog, and have also added more details. Please take a look and let me know if there are more changes required.
cc: @pacoxu